### ISSUE TITLE
Spam reconnection 

### ISSUE THREAD

**Issue by JusteLoneWolf** [discordjs/discord.js#4095]
**Please describe the problem you are having in as much detail as possible:**

Since several weeks (about 3 weeks) my spam bot reconnection to Discord, it happens a few hours or a few minutes after the launch of the shards (Less shards the more it will take time to spam reconnection) I have not made any modification for 4 weeks on my bot so I wonder if it comes from discord directly or from my code.

I have already run it locally and same problem and several of my friends have the same problem on a different library of discord (eris) but also on discord.js 

Namely I use Discord.js 11.6.4 

I currently have 25 servers so I launched my bot plus 25 shards I had no problem for 2-3 minutes then the shards were restarted one after the other. 

Namely that I'm not testing another bot on the same VPS or PC

https://imgur.com/c2Xbf0l.png

**Further details:**

- discord.js version: 11.6.5
- Node.js version: 12.16.1
- Operating system: Linux Debian (VPS)
Windows 10 (Local)
- Priority this issue should have - please be realistic and elaborate if possible: On a scale of 1 to 10 I would say 6/10

Translated with www.DeepL.com/Translator (free version)

**Comment by zechaos031** [comment]
https://imgur.com/tManVca.png another image to illustrate (5 to 10 messages per minute)

**Comment by adrifcastr** [comment]
that is expected behaviour.
Discord psuhed a gateway change that will regularly reconnect you.

**Comment by zechaos031** [comment]
As far as I'm concerned, it's a reconnection every 1 or 2 minutes, I don't think that's really normal.

**Comment by Monbrey** [comment]
You can read more about it from the announcement in the [Discord API Server](https://discordapp.com/channels/81384788765712384/381871767846780928/695380999063863406)

**Comment by vladfrangu** [comment]
Firstly: if you have 25 guilds, you do NOT need 25 shards.. you need 1. You can start sharding when you get to 2.5k servers

Secondly; Discord has now made a change in the gateway API where they will ask connections to reconnect at random times. We implement that accordingly in both v11 and v12 of discord.js, the latter logging when it happens. 

Besides, this isn't really a problem as the bot RESUMED, which doesn't drain the Identify Limit. 

TL;DR: for 25 guilds you don't need to shard to 25 shards, what you've presented is normal, no need to worry about it!

**Comment by zechaos031** [comment]
I know a shards is 2500 guild if I just wanted to see if it was a waiter or something else that was a problem. 

Plus if I want to make shard I do it that way so I don't have to modify all my files once I shard. 

The "it's not really a problem because the bot RESUMED" is sure but during this moment or it is connecting it can take between 3-5 seconds before answering or even not answering. 


But when I'm on a shard it happens after 5 hours of uptime and then it reconnects every 1 minute, so if it was I think it would do it as soon as it was launched.

**Comment by zechaos031** [comment]
![image](https://user-images.githubusercontent.com/36123003/79666279-f81dfd00-81b5-11ea-9991-9315b7996a6d.png)

Are you really sure that's normal?

**Comment by advaith1** [comment]
Jake said bots are now disconnected randomly every 15 minutes to 4 hours, and should automatically resume. The bot shouldn't actually restart, it should just reconnect to the gateway, and as Vlad said, it shouldn't contribute to the connection limit, are you sure there isn't something in your code causing the whole bot to restart?

**Comment by zechaos031** [comment]
Currently I haven't changed the code for about 4 weeks and before that everything was working fine and the bot was reconnecting every 15-20 minutes or so but now it's every minute to 2 and so from one day to the next another person has exactly the same problem as me and he's waiting for a rewrite of the Eris library but he's not sure if it solves the problem.

**Comment by zechaos031** [comment]
Just the number of shards I changed to see if that was the problem but then I went back to 1 shard.

**Comment by JackCrispy** [comment]
still, you don't need 25 shards for 25 servers

**Comment by zechaos031** [comment]
zechaos031 commented 13 hours ago
Just the number of shards I changed to see if that was the problem but then I went back to 1 shard.

**Comment by vladfrangu** [comment]
To make it simpler, can you please:

1. Get some debug logs (listen to the client `debug` event, more info in #3201)
2. Check if the same issue can be seen with v12 of discord.js, with debug logs attached as well as #3201 shows

**Comment by zechaos031** [comment]
I already listen debug event and i ave just this 
![image](https://user-images.githubusercontent.com/36123003/79690674-ea22b780-825b-11ea-84f0-a63bf39f16bc.png)

**Comment by BannerBomb** [comment]
I am commenting this because it seems like a relevant issue to the conversation, but I've noticed this re-connecting thing happening for me randomly too for a few weeks now. My bot only has 1 shard. Most of the time it starts happening as soon as my bot connects to the gateway.
![image](https://user-images.githubusercontent.com/11788894/79712581-37714a00-8299-11ea-89fb-83434830bfcb.png)

And while I'm on this topic, I'm not sure if this is caused from the same issue but I have also noticed that as soon as my bot connects to the gateway a lot of times servers will be reported as unavailable and the guild object name property will just be undefined. It's not like just one server being reported as unavailable. It spams the same servers repeatedly until the bot fully connects to the gateway.
![image](https://user-images.githubusercontent.com/11788894/79712833-fd547800-8299-11ea-865b-51e6c762d7e2.png)

**Comment by AdamMcD94** [comment]
Running into the same issue. The bot runs fine for around an hour or two then begins to disconnect & reconnect continually. 

Only in 1 server and does not use sharding.

DJS Version 11.6.4
Node Version 10.18.1

Seems to have started since that Discord API update referenced before
> You can read more about it from the announcement in the [Discord API Server](https://discordapp.com/channels/81384788765712384/381871767846780928/695380999063863406)

![image](https://user-images.githubusercontent.com/36520109/80125674-e2c81a80-8589-11ea-9994-6528811a327e.png)

**Comment by vladfrangu** [comment]
I don't know if only v11 has this issue or v12 too, can you try it and let me know if it also breaks?

**Comment by zechaos031** [comment]
according to the message above he has the same problem (BannerBomb) because SHARD RECONNECTING, and SHARD RESUME is only on discordjs v12 but I'm gonna try

**Comment by zechaos031** [comment]
miss click sorry

**Comment by zechaos031** [comment]
Hey,

My bot is now offline after 6 days of uptime and I having this error un debug event 
![image](https://user-images.githubusercontent.com/36123003/80321204-854afc80-881b-11ea-88e1-66cfb70f570d.png)

**Comment by zechaos031** [comment]
and I having the last error on my console 

```ErrorEvent {
  target: WebSocket {
    _events: [Object: null prototype] {
      message: [Function],
      open: [Function],
      error: [Function],
      close: [Function]
    },
    _eventsCount: 4,
    _maxListeners: undefined,
    readyState: 2,
    protocol: '',
    _binaryType: 'nodebuffer',
    _closeFrameReceived: false,
    _closeFrameSent: false,
    _closeMessage: '',
    _closeTimer: null,
    _closeCode: 1006,
    _extensions: {},
    _receiver: null,
    _sender: null,
    _socket: null,
    _isServer: false,
    _redirects: 0,
    url: 'wss://gateway.discord.gg/?v=6&encoding=json',
    _req: null,
    [Symbol(kCapture)]: false
  },
  type: 'error',
  message: 'socket hang up',
  error: Error: socket hang up
      at connResetException (internal/errors.js:604:14)
      at TLSSocket.socketOnEnd (_http_client.js:460:23)
      at TLSSocket.emit (events.js:323:22)
      at endReadableNT (_stream_readable.js:1204:12)
      at processTicksAndRejections (internal/process/task_queues.js:84:21) {
    code: 'ECONNRESET'
  }
}
ErrorEvent {
  target: WebSocket {
    _events: [Object: null prototype] {
      message: [Function],
      open: [Function],
      error: [Function],
      close: [Function]
    },
    _eventsCount: 4,
    _maxListeners: undefined,
    readyState: 2,
    protocol: '',
    _binaryType: 'nodebuffer',
    _closeFrameReceived: false,
    _closeFrameSent: false,
    _closeMessage: '',
    _closeTimer: null,
    _closeCode: 1006,
    _extensions: {},
    _receiver: null,
    _sender: null,
    _socket: null,
    _isServer: false,
    _redirects: 0,
    url: 'wss://gateway.discord.gg/?v=6&encoding=json',
    _req: null,
    [Symbol(kCapture)]: false
  },
  type: 'error',
  message: 'socket hang up',
  error: Error: socket hang up
      at connResetException (internal/errors.js:604:14)
      at TLSSocket.socketOnEnd (_http_client.js:460:23)
      at TLSSocket.emit (events.js:323:22)
      at endReadableNT (_stream_readable.js:1204:12)
      at processTicksAndRejections (internal/process/task_queues.js:84:21) {
    code: 'ECONNRESET'
  }
}```

a presicer that I'm using pm2 and that pm2 doesn't detect an error (process id is 0)

![image](https://user-images.githubusercontent.com/36123003/80321293-2e91f280-881c-11ea-87db-6d33b0e49d09.png)

**Comment by vladfrangu** [comment]
Other than BannerBomb who's already on v12, can any/all of you try v12 and report if this is still an issue? Thanks <3

**Comment by zechaos031** [comment]
I'll try that right now. I'll text you back if anything happens.

**Comment by zechaos031** [comment]
Re after feew minutes I have this in logs 
![image](https://user-images.githubusercontent.com/36123003/80323298-e0d0b680-882a-11ea-83e9-41fda83854b1.png) 

in discord.js v12.2.0

**Comment by vladfrangu** [comment]
No issue on our side, it's Discord asking for reconnects and we resume accordingly 🤷‍♂️

**Comment by zechaos031** [comment]
Very strange 😕

**Comment by advaith1** [comment]
It should make you reconnect randomly every 15 mins to 4 hrs

**Comment by AdamMcD94** [comment]
The issue is it seems to be doing this every minute as opposed to every 15. The bot is also responsive during this time. 

As an example, here's the log after the bot has been doing this for 48 hours. The heartbeats are also apparently repeating? It seems more like there's a strange looping issue within the reconnection debug? The fact the bot remains responsive I'm not sure it's even disconnecting/reconnecting when it displays this message. 

![image](https://user-images.githubusercontent.com/36520109/80324087-dd3b3080-8826-11ea-818f-2be7861e5088.png)

**Comment by vladfrangu** [comment]
Alright, so:

1. If you're using v11 and encountering WS issues, upgrade to v12 and report back (after doing the instructions noted in #3201) if the issues persist
2. Discord now sends a RECONNECT Op Code (see https://discordapp.com/developers/docs/topics/gateway#reconnect) at random times. d.js handles this appropriately and attempts a RESUME (see https://discordapp.com/developers/docs/topics/gateway#resuming). Resumes **DO NOT** count towards your identify limit (for 99% of you, said limit is 1000/24 hour), so you're fine. Generally, WS shard closing and resuming is FINE, to be expected and _most of the time_ nothing to worry about! HOWEVER if your shards are sucking up your identify limit like nobody's business, then DEFINITELY report it with attached debug logs

**Comment by AdamMcD94** [comment]
In the process of rewriting to support v12. Once that's all set I'll let you know if it fixes the issue.

Some additional information;

v11 is handling the OpCode 7 (WebSocket close 1001) Reconnect request as expected, however, it begins the odd reconnect debug loop. 

Here is the debug log with some additional logging on the event.

```
[ws] [connection] Heartbeat acknowledged, latency of 110ms
{ t: null, s: null, op: 7, d: null }
[ws] [connection] Attemping to reconnect in 5500ms...
[ws] [connection] Connecting to wss://gateway.discord.gg/?v=6&encoding=json
[ws] [connection] Connected to gateway wss://gateway.discord.gg/?v=6&encoding=json
[ws] [connection] Attempting to resume session a1439a15ac571117aa42bf1c7b5ba3af
[ws] [connection] Setting a heartbeat interval for 41250ms
[ws] [connection] RESUMED | replayed 93 events.
[ws] [connection] Sending a heartbeat
[ws] [connection] Heartbeat acknowledged, latency of 114ms
CloseEvent {
  target:
   WebSocket {
     _events:
      [Object: null prototype] {
        message: [Function],
        open: [Function],
        error: [Function],
        close: [Function] },
     _eventsCount: 4,
     _maxListeners: undefined,
     readyState: 3,
     protocol: '',
     _binaryType: 'nodebuffer',
     _closeFrameReceived: true,
     _closeFrameSent: true,
     _closeMessage: 'Discord WebSocket requesting client reconnect.',
     _closeTimer:
      Timeout {
        _called: false,
        _idleTimeout: -1,
        _idlePrev: null,
        _idleNext: null,
        _idleStart: 4898939,
        _onTimeout: null,
        _timerArgs: undefined,
        _repeat: null,
        _destroyed: false,
        [Symbol(unrefed)]: false,
        [Symbol(asyncId)]: 1014,
        [Symbol(triggerId)]: 49 },
     _closeCode: 1001,
     _extensions: {},
     _receiver:
      Receiver {
        _writableState: [WritableState],
        writable: false,
        _events: [Object: null prototype] {},
        _eventsCount: 0,
        _maxListeners: undefined,
        _binaryType: 'nodebuffer',
        _extensions: {},
        _maxPayload: 104857600,
        _bufferedBytes: 0,
        _buffers: [],
        _compressed: false,
        _payloadLength: 48,
        _mask: undefined,
        _fragmented: 0,
        _masked: false,
        _fin: true,
        _opcode: 8,
        _totalPayloadLength: 0,
        _messageLength: 0,
        _fragments: [],
        _state: 0,
        _loop: false,
        [Symbol(websocket)]: [Circular] },
     _sender:
      Sender {
        _extensions: {},
        _socket: [TLSSocket],
        _firstFragment: true,
        _compress: false,
        _bufferedBytes: 0,
        _deflating: false,
        _queue: [] },
     _socket:
      TLSSocket {
        _tlsOptions: [Object],
        _secureEstablished: true,
        _securePending: false,
        _newSessionPending: false,
        _controlReleased: true,
        _SNICallback: null,
        servername: 'gateway.discord.gg',
        alpnProtocol: false,
        authorized: true,
        authorizationError: null,
        encrypted: true,
        _events: [Object],
        _eventsCount: 4,
        connecting: false,
        _hadError: false,
        _handle: null,
        _parent: null,
        _host: 'gateway.discord.gg',
        _readableState: [ReadableState],
        readable: false,
        _maxListeners: undefined,
        _writableState: [WritableState],
        writable: false,
        allowHalfOpen: false,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: undefined,
        _server: null,
        ssl: null,
        _requestCert: true,
        _rejectUnauthorized: true,
        parser: null,
        _httpMessage: null,
        timeout: 0,
        write: [Function: writeAfterFIN],
        [Symbol(res)]: null,
        [Symbol(asyncId)]: 49,
        [Symbol(lastWriteQueueSize)]: 0,
        [Symbol(timeout)]: null,
        [Symbol(kBytesRead)]: 71917,
        [Symbol(kBytesWritten)]: 3134,
        [Symbol(connect-options)]: [Object],
        [Symbol(websocket)]: undefined },
     _isServer: false,
     _redirects: 0,
     url: 'wss://gateway.discord.gg/?v=6&encoding=json',
     _req: null },
  type: 'close',
  wasClean: true,
  reason: 'Discord WebSocket requesting client reconnect.',
  code: 1001 }
[ws] [connection] Server closed the WebSocket connection: 1001
[ws] [connection] Clearing heartbeat interval
[ws] [connection] Attemping to reconnect in 5500ms...
[ws] [connection] Connecting to wss://gateway.discord.gg/?v=6&encoding=json
[ws] [connection] Connected to gateway wss://gateway.discord.gg/?v=6&encoding=json
[ws] [connection] Attempting to resume session a1439a15ac571117aa42bf1c7b5ba3af
[ws] [connection] Setting a heartbeat interval for 41250ms
[ws] [connection] RESUMED | replayed 1 events.
[ws] [connection] Sending a heartbeat
[ws] [connection] Heartbeat acknowledged, latency of 110ms
[ws] [connection] Sending a heartbeat
[ws] [connection] Heartbeat acknowledged, latency of 112ms
[ws] [connection] Sending a heartbeat
[ws] [connection] Heartbeat acknowledged, latency of 111ms
CloseEvent {
  target:
   WebSocket {
     _events:
      [Object: null prototype] {
        message: [Function],
        open: [Function],
        error: [Function],
        close: [Function] },
     _eventsCount: 4,
     _maxListeners: undefined,
     readyState: 3,
     protocol: '',
     _binaryType: 'nodebuffer',
     _closeFrameReceived: true,
     _closeFrameSent: true,
     _closeMessage: '',
     _closeTimer:
      Timeout {
        _called: false,
        _idleTimeout: -1,
        _idlePrev: null,
        _idleNext: null,
        _idleStart: 4950242,
        _onTimeout: null,
        _timerArgs: undefined,
        _repeat: null,
        _destroyed: false,
        [Symbol(unrefed)]: false,
        [Symbol(asyncId)]: 1071,
        [Symbol(triggerId)]: 987 },
     _closeCode: 1000,
     _extensions: {},
     _receiver:
      Receiver {
        _writableState: [WritableState],
        writable: false,
        _events: [Object: null prototype] {},
        _eventsCount: 0,
        _maxListeners: undefined,
        _binaryType: 'nodebuffer',
        _extensions: {},
        _maxPayload: 104857600,
        _bufferedBytes: 0,
        _buffers: [],
        _compressed: false,
        _payloadLength: 2,
        _mask: undefined,
        _fragmented: 0,
        _masked: false,
        _fin: true,
        _opcode: 8,
        _totalPayloadLength: 0,
        _messageLength: 0,
        _fragments: [],
        _state: 0,
        _loop: false,
        [Symbol(websocket)]: [Circular] },
     _sender:
      Sender {
        _extensions: {},
        _socket: [TLSSocket],
        _firstFragment: true,
        _compress: false,
        _bufferedBytes: 0,
        _deflating: false,
        _queue: [] },
     _socket:
      TLSSocket {
        _tlsOptions: [Object],
        _secureEstablished: true,
        _securePending: false,
        _newSessionPending: false,
        _controlReleased: true,
        _SNICallback: null,
        servername: 'gateway.discord.gg',
        alpnProtocol: false,
        authorized: true,
        authorizationError: null,
        encrypted: true,
        _events: [Object],
        _eventsCount: 4,
        connecting: false,
        _hadError: false,
        _handle: null,
        _parent: null,
        _host: 'gateway.discord.gg',
        _readableState: [ReadableState],
        readable: false,
        _maxListeners: undefined,
        _writableState: [WritableState],
        writable: false,
        allowHalfOpen: false,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: undefined,
        _server: null,
        ssl: null,
        _requestCert: true,
        _rejectUnauthorized: true,
        parser: null,
        _httpMessage: null,
        timeout: 0,
        write: [Function: writeAfterFIN],
        [Symbol(res)]: null,
        [Symbol(asyncId)]: 987,
        [Symbol(lastWriteQueueSize)]: 0,
        [Symbol(timeout)]: null,
        [Symbol(kBytesRead)]: 976,
        [Symbol(kBytesWritten)]: 423,
        [Symbol(connect-options)]: [Object],
        [Symbol(websocket)]: undefined },
     _isServer: false,
     _redirects: 0,
     url: 'wss://gateway.discord.gg/?v=6&encoding=json',
     _req: null },
  type: 'close',
  wasClean: true,
  reason: '',
  code: 1000 }
[ws] [connection] Server closed the WebSocket connection: 1000
[ws] [connection] Clearing heartbeat interval
[ws] [connection] Attemping to reconnect in 5500ms...
[ws] [connection] Sending a heartbeat
[ws] [connection] Heartbeat acknowledged, latency of 117ms
[ws] [connection] Connecting to wss://gateway.discord.gg/?v=6&encoding=json
[ws] [connection] Connected to gateway wss://gateway.discord.gg/?v=6&encoding=json
[ws] [connection] Attempting to resume session a1439a15ac571117aa42bf1c7b5ba3af
[ws] [connection] Setting a heartbeat interval for 41250ms
[ws] [connection] RESUMED | replayed 2 events.
```

It continues to display code 1000 connection resets every 3 heartbeat intervals.

Possible theory - When djs receives Opcode 7 it begins the reconnection. Once reconnected it replays events which includes a close event (the expected 1001 from Discord requesting the reconnection) which also begins another reconnection attempt.

**Comment by azbaron** [comment]
Seeing the same shard reconnects when using internal sharding on v12.
Why would Discord change to request a reconnect every so often? Finding it very obnoxious, as i'm running 3 shards (on 2,700+ guilds) and it usually spams a reconnect on a shard every 3-5 minutes. 

It wouldn't bother me if it wasn't delaying bot responses for users.

**Comment by vladfrangu** [comment]
> Why would Discord change to request a reconnect every so often? 

![image](https://user-images.githubusercontent.com/17960496/80417693-d112aa00-88de-11ea-8c23-42c457e2a447.png)

Nothing we can do other than close and resume as the docs tell us to. Sorry, but at this point, this is out of our hands (unless it matches anything said in https://github.com/discordjs/discord.js/issues/4095#issuecomment-619652466)

**Comment by advaith1** [comment]
![image](https://advaith.is-in-hyper.space/b319d00f84.png)

it shouldn't be every 3-5 minutes though 🤔

**Comment by vladfrangu** [comment]
Again: I don't know if v11 is broken in this regard. However, v12 clearly logs when Discord asks us for a reconnect as seen in comment https://github.com/discordjs/discord.js/issues/4095#issuecomment-619645735
![image](https://user-images.githubusercontent.com/17960496/80425009-83506e80-88eb-11ea-93ea-05fa6723e239.png)

This is **not** a library bug and this is out of our hands. 
![image](https://user-images.githubusercontent.com/17960496/80425204-e215e800-88eb-11ea-96ae-53e9ed0cced8.png)

If you get IDENTIFY's every 3-5 minutes..that's a massive issue and DO report it.

---

However! I have gotten in touch with Mason, and he said that, and I quote, `If he’s seeing spam reconnects every 3 minutes on one shard I can escalate`. As such: @azbaron, @AdamMcD94, @zechaos031 (and anyone else reading this), please follow these steps (more-less) and add debug logging, followed by DMing me (Vladdy#0002) on Discord once you're starting to encounter the spam resumes and we'll get this solved!

##### Example code:

```js
const { createWriteStream, unlinkSync } = require('fs');

try { unlinkSync('./gateway-logging.log') } catch {}

const stream = createWriteStream('./gateway-logging.log');

// Add this code BEFORE logging in
client.on('raw', ({ t, d }, id) => {
	const timestamp = new Date().toISOString();
	switch (t) {
		case 'READY': {
			console.log(`[${timestamp}] Shard ${id} READY | Trace: ${d._trace}`);
			stream.write(`[${timestamp}] Shard ${id} READY | Trace: ${d._trace}`);
			break;
		}
		case 'RESUMED': {
			console.log(`[${timestamp}] Shard ${id} RESUMED | Trace: ${d._trace}`);
			stream.write(`[${timestamp}] Shard ${id} RESUMED | Trace: ${d._trace}`);
			break;
		}
	}
});

client.on('debug', message => {
	// Ignore spammy messages
	if (/(Sending a heartbeat|Latency of)/i.test(message)) return null;
	if (/voice/i.test(message)) return null;

	const timestamp = new Date().toISOString();
	console.log(`[${timestamp}] ${message}`);
	stream.write(`[${timestamp}] ${message}\n`);
});
```

**Comment by zechaos031** [comment]
I adding this code to my bot and I waitings ✔️

**Comment by vladfrangu** [comment]
@zechaos031 just a note: v11 most likely has a bug related to this whole thing, which I don't know if it will get fixed (citation needed). Make sure you try this on v12 (same applies for everyone else)

**Comment by zechaos031** [comment]
yes this code instantly crashed my bot x) 

I started it in v12 right now

**Comment by AdamMcD94** [comment]
Already discussed with vladfrangu on Discord but to update here, after 72 hours I've yet to encounter the same issue on v12. Seems to have been solved by updating.

**Comment by azbaron** [comment]
Just to note: I am on V12.2.0 and am still having the shard reconnect spams. I have already sent logs to Vlad and have extra logs on hand if needed _(had it happen to me really good yesterday)._

The issue doesn't occur all the time, as I do see normal re-connects like vlad described as being normal behavior; however, i'll randomly have one of my shards do it for a solid 1-5 minutes straight.

**Comment by vladfrangu** [comment]
@azbaron mind sharing those logs to me again on Discord? I'll check them out later today (when it's morning and not 4:32 in the morning)!

**Comment by azbaron** [comment]
> @azbaron mind sharing those logs to me again on Discord? I'll check them out later today (when it's morning and not 4:32 in the morning)!
Done

**Comment by azbaron** [comment]
Bringing this back up. Updated to 12.4.1 and am now hitting the rate limit cap for sessionStartLimit (no unhandled restarts at all, only 5-10 manual reboots within 24 hours).
Going to try to get more logs and upload here.

**Comment by azbaron** [comment]
@vladfrangu I have recent logs, may I send them to you?

**Comment by vladfrangu** [comment]
> @vladfrangu I have recent logs, may I send them to you?

Of course!

**Comment by azbaron** [comment]
Vladd and I have discussed and cannot figure out why the shards are constantly reconnecting.

I have been doing internal sharding. I switched over to using ShardingManger and seem to have a bit more stability. Which leaves me to believe there is a bug with internal sharding, or at least, when it gets over 2 shards.

**Comment by cobsidian-crim** [comment]
still happening to me and i use Sharding Manager

**Comment by JMTK** [comment]
I recently have run into this where the spam reconnection happens a lot, and then sometimes can get "corrupted" where it never will reconnect to the gateway until I restart the process and presumably get a new session ID. The message that appears when this happens is:
```
[WS => Shard 1] Failed to connect to the gateway, requeueing...
```

I am not sure what other code samples I can provide besides whats been given but I can happily add more logs

**Comment by kyranet** [comment]
Does this issue still happen in v13? Are you using internal sharding, or process-based sharding (ShardingManager)?

**Comment by JMTK** [comment]
> Does this issue still happen in v13? Are you using internal sharding, or process-based sharding (ShardingManager)?

I would say it's happening a lot less now on v13 but I have had it stay in a "non-ready" state for extended periods(5-10+ minutes) before I restart the instance. I use internal sharding.

**Comment by azbaron** [comment]
> Does this issue still happen in v13? Are you using internal sharding, or process-based sharding (ShardingManager)?

I haven't had the issue so far in v13. I actually haven't seen the issue in a while; although with v12, my client did seem to refuse to get stuck in a non-ready state, similar to what JMTK mentioned. This ended up forcing me to rush update to v13 faster than I had planned (not a bad thing, really).
Sharding has actually seemed to be a lot more stable so far with v13.x. Although, I also haven't published my latest update with v13 to my main client yet...which is usually running about 7 shards.

**Comment by JMTK** [comment]
Happened again today. I received about 4500 of this error from the `shardError` event before I just force restarted it:
```{"message":"Zero length buffer.","stack":"Error: Zero length buffer.
    at Object.unpack (<anonymous>)
    at Object.exports.unpack (/app/node_modules/discord.js/src/WebSocket.js:26:18)
    at WebSocketShard.onMessage (/app/node_modules/discord.js/src/client/websocket/WebSocketShard.js:293:26)
    at WebSocket.onMessage (/app/node_modules/ws/lib/event-target.js:199:18)
    at WebSocket.emit (node:events:390:28)
    at Receiver.receiverOnMessage (/app/node_modules/ws/lib/websocket.js:1022:20)
    at Receiver.emit (node:events:390:28)
    at Receiver.dataMessage (/app/node_modules/ws/lib/receiver.js:507:14)
    at Receiver.getData (/app/node_modules/ws/lib/receiver.js:440:17)
    at Receiver.startLoop (/app/node_modules/ws/lib/receiver.js:148:22)","name":"Error"}```

**Comment by kyranet** [comment]
Are you using Erlpack? What happens if you use JSON instead?

**Comment by louisgv** [comment]
I'm receiving Discord's system notification on the 1000s message limit as well. Below are some logs:
```
[2021-12-21T13:57:22.365Z] [WS => Shard 0] Setting a HELLO timeout for 20s.                                                         
[2021-12-21T13:57:22.722Z] [WS => Shard 0] [CONNECTED] Took 356ms [2021-12-21T13:57:22.724Z] [WS => Shard 0] Clearing the HELLO time out
[2021-12-21T13:57:22.724Z] [WS => Shard 0] Setting a heartbeat interval for 41250ms.                                                
[2021-12-21T13:57:22.726Z] [WS => Shard 0] [IDENTIFY] Shard 0/1 with intents: 4609                                                  
[2021-12-21T13:57:22.978Z] [WS => Shard 0] [READY] Session 963e52f1309f11687e9958dc6b6c7d30.
[2021-12-21T13:57:22.986Z] [WS => Shard 0] Shard received all its guilds. Marking as fully ready.
[2021-12-21T13:58:18.507Z] [WS => Shard 0] [CLOSE]
    Event Code: 1000                                                  
    Clean     : true
    Reason    :
[2021-12-21T13:58:18.508Z] [WS => Shard 0] Clearing the heartbeat interval.                                                         
[2021-12-21T13:58:18.508Z] [WS => Shard 0] WS State: CLOSED
[2021-12-21T13:58:18.509Z] [WS => Shard 0] [CONNECT]                  
    Gateway    : wss://gateway.discord.gg/
    Version    : 9
    Encoding   : json
    Compression: none
[2021-12-21T13:58:18.509Z] [WS => Shard 0] Setting a HELLO timeout for 20s.                                                         
[2021-12-21T13:58:18.947Z] [WS => Shard 0] [CONNECTED] Took 438ms
[2021-12-21T13:58:18.949Z] [WS => Shard 0] Clearing the HELLO timeout.                                                              
[2021-12-21T13:58:18.949Z] [WS => Shard 0] Setting a heartbeat interval for 41250ms.                                                
[2021-12-21T13:58:18.950Z] [WS => Shard 0] [IDENTIFY] Shard 0/1 with intents: 4609                                          
[2021-12-21T13:58:19.520Z] [WS => Shard 0] [CLOSE]                    
Event Code: 4004                                                  
Clean     : true                                                  
Reason    : Authentication failed.                            
[2021-12-21T13:58:19.522Z] [WS => Shard 0] Clearing the heartbeat interval.                                                         
[2021-12-21T13:58:19.527Z] [WS =>c  Shard 0] TOKEN_INVALID          
[2021-12-21T13:58:19.536Z] [WS => Manager] Couldn't reconnect or fetch information about the gateway. Error [TOKEN_INVALID]: An invalid token was provided.                                           
[2021-12-21T13:58:19.537Z] [WS => Manager] Possible network error occurred. Retrying in 5s...
```

^ As can be seen, the WS connection connect and close almost every minute??

**Comment by JMTK** [comment]
> erlpack v0.1.3
> 
> I can try without it and report back!

I know this is much later but I have been on v14 for a while now and while I haven't gotten that "Zero length buffer" error, I have this event from `shardError` occasionally now, and I will receive it all as a short burst(100s - 1000s at a time)
```
TypeError [ERR_INVALID_ARG_TYPE]: The "input" argument must be an instance of ArrayBuffer or ArrayBufferView. Received null
 at new NodeError (node:internal/errors:387:5)
 at TextDecoder.decode (node:internal/encoding:417:15)
 at exports.unpack (/app/node_modules/discord.js/src/WebSocket.js:22:17)
 at WebSocketShard.onMessage (/app/node_modules/discord.js/src/client/websocket/WebSocketShard.js:314:26)
 at WebSocket.onMessage (/app/node_modules/ws/lib/event-target.js:199:18)
 at WebSocket.emit (node:events:513:28)
 at Receiver.receiverOnMessage (/app/node_modules/ws/lib/websocket.js:1178:20)
 at Receiver.emit (node:events:513:28)
 at Receiver.dataMessage (/app/node_modules/ws/lib/receiver.js:513:14)
 at Receiver.getData (/app/node_modules/ws/lib/receiver.js:446:17)
 ```
 
 I'm unsure on how to determine what data it's actually receiving, but when that happens then I notice that the shard permanently will stay stuck on a certain status, like `Identifying (7)` and essentially just won't ever recover and I just restart the process

**Comment by Jiralite** [comment]
Closing this as it is a stale issue. Feel free to make a new issue if this problem still persists.
